pgjdbc/src/main/java/org/postgresql/ssl/SSLFactory.java

```diff
@@ -0,0 +1,195 @@
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.util.Properties;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509KeyManager;
+import javax.security.auth.x500.X500Principal;
+
+/**
+ * Socket factory that uses typical "zero configuration" key and trust
+ * stores, like the native Windows "SunMSCAPI" or native Apple "Apple"
+ * providers.
+ */
+public class SSLFactory extends WrappedFactory {
+
+  protected SSLFactory(Properties info, final String protocol, final String algorithm, final String keyStoreProvider, final String keyStoreType, final String trustStoreProvider, final String trustStoreType) throws PSQLException {
+
+    SSLContext ctx;
+    final KeyStore keyStore;
+    KeyStore trustStore;
+    final char[] keyPassphrase = null;
+    final String sslsubject = PGProperty.SSL_SUBJECT.getOrDefault(info);
+    final X500Principal subject;
+    final KeyManagerFactory keyManagerFactory;
+    final TrustManagerFactory trustManagerFactory;
+
+    try {
+
+      keyStore = KeyStore.getInstance(keyStoreType, keyStoreProvider);
+
+    } catch (KeyStoreException ex) {
+      throw new PSQLException(GT.tr("SSL keystore {0} not available.",
+                keyStoreType), PSQLState.CONNECTION_FAILURE, ex);
+    } catch (NoSuchProviderException ex) {
+      throw new PSQLException(GT.tr("SSL keystore {0} not available.",
+                keyStoreType), PSQLState.CONNECTION_FAILURE, ex);
+    }
+
+    try {
+
+      keyStore.load(null, null);
+
+    } catch (CertificateException ex) {
+      throw new PSQLException(GT.tr("SSL keystore {0} could not be loaded.",
+                keyStoreType), PSQLState.CONNECTION_FAILURE, ex);
+    } catch (IOException ex) {
+      throw new PSQLException(GT.tr("SSL keystore {0} could not be loaded.",
+                keyStoreType), PSQLState.CONNECTION_FAILURE, ex);
+    } catch (NoSuchAlgorithmException ex) {
+      throw new PSQLException(GT.tr("Could not find a java cryptographic algorithm: {0}.",
+                ex.getMessage()), PSQLState.CONNECTION_FAILURE, ex);
+    }
+
+    try {
+
+      keyManagerFactory = KeyManagerFactory
+              .getInstance(algorithm);
+
+    } catch (NoSuchAlgorithmException ex) {
+      throw new PSQLException(GT.tr("Could not find a java cryptographic algorithm: {0}.",
+                ex.getMessage()), PSQLState.CONNECTION_FAILURE, ex);
+    }
+
+    try {
+
+      keyManagerFactory.init(keyStore, keyPassphrase);
+
+    } catch (NoSuchAlgorithmException ex) {
+      throw new PSQLException(GT.tr("Could not find a java cryptographic algorithm: {0}.",
+                ex.getMessage()), PSQLState.CONNECTION_FAILURE, ex);
```

Comment:
Isn't `ex.getMessage()` redudant since the target message will be recursively built?

Reply:
Java is fixed.

The ex.getMessage() is there on purpose so that the whole error message is on one single line.

The target audience for this patch are data scientists who aren't Java developers, nor are they experts in SSL. They see the message "Could not find a java cryptographic algorithm", they don't understand it (fair: "which crypto algorithm?"), and I have to pick it all apart for them, first helping them give me the whole exception (I'll get a screenshot of the top few lines, then I'll explain how to cut and paste), then finding the needle in the haystack of "caused by", then googling for them.

With the whole message on one line, they have one line to google themselves, on the top line, and lots and lots of make work is avoided.